### PR TITLE
[SEMI-MODULAR] Allows observers the ability to activate pAI cards on their own

### PR DESCRIPTION
**Pull request by Bobbanz1**
## About The Pull Request
What this PR does is that it allows observing players the ability to join the round as a pAI without having someone in-round be needed to activate a pAI card, wait for someone to submit a personality and then download said personality.

* Closes https://github.com/Bubberstation/Bubberstation/issues/140

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>

https://user-images.githubusercontent.com/59128051/235649357-54772d46-aae6-401d-b565-066fcb92da69.mp4

</details>

## Changelog

:cl:
add: Grants observers the ability to activate and possess pAI cards without requiring someone to be on hand to activate the pAI card for them.
/:cl:

### CODE REVIEW

**Comment by Bobbanz1** [comment]
Now why did you fail to go through the checks...

**Comment by Cyprex** [comment]
CI failure isn't your fault

File: modular_zubbers/code/modules/pai/ghost_pai.dm [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/obj/item/pai_card/attack_ghost(mob/user)
+	ghost_activate(user)
```
Comment (Cyprex) [inline modular_zubbers/code/modules/pai/ghost_pai.dm:2]: call parent (`. = ..()`), probably before ghost_activate?

File: modular_zubbers/code/modules/pai/ghost_pai.dm [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Checks if the user is allowed to be a pAI, and if so, creates a new pAI mob.
+ * Assuming the pAI card itself is empty of a personality.
+ */
+/obj/item/pai_card/proc/ghost_activate(mob/user)
+	if(pai)
+		return FALSE
```
Comment (Cyprex) [inline modular_zubbers/code/modules/pai/ghost_pai.dm:10]:
```suggestion
		return
```
The return value is never checked anywhere

File: modular_zubbers/code/modules/pai/ghost_pai.dm [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+		to_chat(user, span_warning("Personality Matrix Corrupted! Please reload Personality file (Open the pAI submit window and click the load button before trying again)"))
+		return
+	var/datum/pai_candidate/candidate = SSpai.candidates[pai_ckey]
+	var/mob/living/silicon/pai/ghost_pai = new(src)
+	ghost_pai.name = candidate.name || pick(GLOB.ninja_names)
+	ghost_pai.real_name = ghost_pai.name
+	ghost_pai.key = candidate.ckey
+	set_personality(ghost_pai)
+	SSpai.candidates -= pai_ckey
+	return TRUE
```
Comment (Cyprex) [inline modular_zubbers/code/modules/pai/ghost_pai.dm:27]:
```suggestion
	return
```

File: modular_zubbers/code/modules/pai/ghost_pai.dm [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ */
+/obj/item/pai_card/proc/ghost_activate(mob/user)
+	if(pai)
+		return FALSE
+	var/pai_ask = tgui_alert(user, "Become a pAI? (Warning, You can no longer be revived, and all past lives will be forgotten!)", "Confirm", list("Yes","No"))
+	if(pai_ask != "Yes" || QDELETED(src))
+		return
+	var/pai_ckey = user.ckey
+	if(!user.client || !isobserver(user) || is_banned_from(pai_ckey, ROLE_PAI))
+		return FALSE
```
Comment (Cyprex) [inline modular_zubbers/code/modules/pai/ghost_pai.dm:16]:
```suggestion
		return
```

File: modular_zubbers/code/modules/pai/ghost_pai.dm [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	if(!SSpai.candidates[pai_ckey])
+		to_chat(user, span_warning("Personality Matrix Corrupted! Please reload Personality file (Open the pAI submit window and click the load button before trying again)"))
```
Comment (Cyprex) [inline modular_zubbers/code/modules/pai/ghost_pai.dm:18]: errr, is this the error message for when someone clicked yes but doesn't have PAI settings set up?

**Review by Cyprex — CHANGES_REQUESTED** [review]
I'm curious- could you rely on download_candidate instead of replicating most of its functionality here?

File: modular_zubbers/code/modules/pai/ghost_pai.dm (hunk 5 above)
Comment (Bobbanz1) [reply modular_zubbers/code/modules/pai/ghost_pai.dm:18]: > errr, is this the error message for when someone clicked yes but doesn't have PAI settings set up?

Aye

**Comment by Bobbanz1** [comment]
> I'm curious- could you rely on download_candidate instead of replicating most of its functionality here?

Hmmm I can't use that proc as the download_candidate proc runs through the candidate check_ready proc which checks if the player has submitted a personality beforehand to be available for downloading...hmmm it could also be possible that I make it so you have to have submitted a personality in order to download as a ghost, which would remove the need for the error message.

**Comment by Cyprex** [comment]
Good to merge after testing

**Comment by InterroLouis** [comment]
Seeing as this is still not merged, I don't suppose it would be possible for you, the coder responsible for this PR, to also map in some PAI card spawn points somewhere other than the Research Director's Office so that people might use them?

**Comment by Bobbanz1** [comment]
> Seeing as this is still not merged, I don't suppose it would be possible for you, the coder responsible for this PR, to also map in some PAI card spawn points somewhere other than the Research Director's Office so that people might use them?

Pretty certain by default at least 1 PAI card spawns in the dormitory locker room area but I can add them to more places

**Comment by InterroLouis** [comment]
Various places? One spawning is nice but that's basically just a single PAI player, or one person just taking the card and shoving it in their satchel or webbing. It would be nice to see more spawn points in various places. The specific locations don't matter as much as that they exist at all, although my perhaps unwise ideal would be a spot where the PAI can get out from but there's much less chance of people simply picking up the card and putting it in their inventory.

**Comment by Bobbanz1** [comment]
> Various places? One spawning is nice but that's basically just a single PAI player, or one person just taking the card and shoving it in their satchel or webbing. It would be nice to see more spawn points in various places. The specific locations don't matter as much as that they exist at all, although my perhaps unwise ideal would be a spot where the PAI can get out from but there's much less chance of people simply picking up the card and putting it in their inventory.

I've added 2 more PAI cards to each currently available public spawn site so there's in total 3 PAI cards on each site, I haven't put any on Birdshot or Northstar as I don't know if those maps are even used

**Comment by nevimer** [comment]
Make sure that you add a check to have people not spawn before the round starts! It's currently possible, not a big deal.

**Comment by Bobbanz1** [comment]
> Make sure that you add a check to have people not spawn before the round starts! It's currently possible, not a big deal.

Did so now!

**Comment by nevimer** [comment]
Update from master, I want to merge this fully.

**Pull request merged**